src/class-big-pit.php

```diff
@@ -0,0 +1,219 @@
+ */
+final class Big_Pit implements Feature {
+	/**
+	 * Singleton.
+	 *
+	 * @var self
+	 */
+	private static ?self $instance = null;
+
+	/**
+	 * Whether the database is ready to accept queries.
+	 *
+	 * @var bool
+	 */
+	private bool $ready = false;
+
+	/**
+	 * Instance.
+	 *
+	 * @return self
+	 */
+	public static function instance(): self {
+		if ( null === self::$instance ) {
+			self::$instance = new self();
+			self::$instance->boot();
+		}
+
+		return self::$instance;
+	}
+
+	/**
+	 * Boot the feature.
+	 */
+	public function boot(): void {
+		global $wpdb;
+
+		$wpdb->big_pit = $wpdb->base_prefix . 'big_pit';
+
+		if ( defined( 'WP_INSTALLING' ) ) {
+			return;
+		}
+
+		$this->upsert();
+		$this->ready = true;
+	}
+
+	/**
+	 * Get a value.
+	 *
+	 * @param string $key   Item key.
+	 * @param string $group Item group.
+	 * @return mixed|null
+	 */
+	public function get( string $key, string $group ): mixed {
+		global $wpdb;
+
+		if ( ! $this->ready ) {
+			return null;
+		}
+
+		$value = $wpdb->get_var(
+			$wpdb->prepare(
+				"SELECT item_value FROM {$wpdb->big_pit} WHERE item_group = %s AND item_key = %s LIMIT 1",
+				$group,
+				$key
+			),
+		);
+
+		if ( is_string( $value ) ) {
+			$value = maybe_unserialize( $value );
+		}
+
+		return $value;
+	}
+
+	/**
+	 * Set a value.
+	 *
+	 * @param string $key   Item key.
+	 * @param mixed  $value Item value.
+	 * @param string $group Item group.
+	 */
+	public function set( string $key, mixed $value, string $group ): void {
+		global $wpdb;
+
+		if ( ! $this->ready ) {
+			return;
+		}
+
+		$value = maybe_serialize( $value ); // @phpstan-ignore argument.type
+
+		$exists = $wpdb->get_var(
+			$wpdb->prepare(
+				"SELECT item_id FROM {$wpdb->big_pit} WHERE item_group = %s AND item_key = %s LIMIT 1",
+				$group,
+				$key
+			),
+		);
+
+		if ( $exists ) {
+			$wpdb->update(
+				$wpdb->big_pit,
+				[
+					'item_value' => $value,
+				],
+				[
+					'item_id' => $exists,
+				],
+				[ '%s' ],
+				[ '%s' ],
+			);
+		} else {
+			$wpdb->insert(
+				$wpdb->big_pit,
+				[
+					'item_group' => $group,
+					'item_key'   => $key,
+					'item_value' => $value,
+				],
+				[ '%s', '%s', '%s' ],
+			);
+		}
+	}
+
+	/**
+	 * Delete a value.
+	 *
+	 * @param string $key   Item key.
+	 * @param string $group Item group.
+	 */
+	public function delete( string $key, string $group ): void {
+		global $wpdb;
+
+		if ( ! $this->ready ) {
+			return;
+		}
+
+		$wpdb->delete(
+			$wpdb->big_pit,
+			[
+				'item_group' => $group,
+				'item_key'   => $key,
+			],
+			[ '%s', '%s' ],
+		);
+	}
+
+	/**
+	 * Delete all values in a group.
+	 *
+	 * @param string $group Item group.
+	 */
+	public function flush_group( string $group ): void {
+		global $wpdb;
+
+		if ( ! $this->ready ) {
+			return;
+		}
+
+		$wpdb->delete(
+			$wpdb->big_pit,
+			[
+				'item_group' => $group,
+			],
+			[ '%s' ],
+		);
+	}
+
+	/**
+	 * Create or update the database table.
+	 */
+	private function upsert(): void {
```

Comment:
Any value in returning `bool` here, `true` if it was already ready or if the db delta completed, `false` if not? You could check that result before declaring the featured `ready`.